ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -796,27 +843,35 @@ module AuthADWinbind : Auth_signature.AUTH_MODULE = struct
     ; ("subject-is-group", string_of_bool true)
     ]
 
-  let query_subject_information_user (name : string) (uid : int) (sid : string)
-      =
-    let* {gecos; gid} = Wbinfo.uid_info_of_uid uid in
+  let query_subject_information_user (uid : int) (sid : string) =
+    let* {user_name; gecos; gid} = Wbinfo.uid_info_of_uid uid in
+    let* domain = Wbinfo.domain_of_uname user_name in
     let* {
-           upn
+           name
+         ; upn
+         ; display_name
          ; account_disabled
          ; account_expired
          ; account_locked
          ; password_expired
          } =
-      Ldap.query_user sid
+      Ldap.query_user sid domain
     in
     Ok
       [
-        ("subject-name", name)
+        ("subject-name", user_name)
       ; ("subject-gecos", gecos)
       ; ( "subject-displayname"
-        , if gecos = "" || gecos = "<null>" then name else gecos )
+        , if display_name != "" then
+            display_name
+          else if gecos != "" && gecos != "<null>" then
+            gecos
+          else
+            user_name )
```

Comment:
you have to be careful in ocaml because `!=` does not do what you might expect (see `!=` vs `<>` https://stackoverflow.com/a/1412688)

please do the following replacement `s/!=/<>/g`

Reply:
nice catch!